_published/0012-authentication.md

```diff
@@ -0,0 +1,167 @@
+
+# Update the following (it's YAML syntax)
+
+pr_id: 12 # Update this with PR number/ID. No leading zeros rfc_feature_name:
+authentication # Use kebab-case title: "RFC-0012: Authentication" # Update this
+with PR number/ID and feature name. Use leading zeros rfc_author_username:
+fredx30 rfc_author_name: Fredrik Dyrvold # Or same as username, if you wish
+
+# Leave these. Collaborator changes this before merging
+
+impl_issue_id: 0 impl_issue_repo: iver-wharf/wharf-api last_modified_date:
+2021-06-16
+---
+
+# {{page.title}}
+
+- RFC
+  PR: [iver-wharf/rfcs#{{page.pr_id}}](https://github.com/iver-wharf/rfcs/pulls/{{page.pr_id}})
+- Feature name: `{{page.rfc_feature_name}}`
+- Author:
+  {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation
+  issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation
+  status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Wharf should have authentication. Wharf currently has no authentication. This
+will outline a plan to implement [OpenID Connect (OICD)](https://openid.net/connect/)
+in wharf. This RFC will only outline authentication with OICD and not
+authorization and user/group
+management. A further plan to implement user groups and authorization is
+expected to be drawn up in a later document.
+
+## Motivation
+
+Some claim a good feature of secure infrastructure is access control. Currently,
+our only solution is network isolation. Being able to restrict access to varying
+degrees is a product goal to satisfy business needs.
+
+## Explanation
+
+Pre implementation stages
+
+1. OICD in any form demands full SSL security in all connections. In order to
+   avoid having to fill the development environment with --force-unsecure
+   bindings we should consider implementing certificate generation and importing
+   for dev environments.
+
+Implementations stages
+
+1. Docker compose setup for Hydra Proxy setup
+   - This involves integration with the currently existing DB.
+2. Hydra (Authorisation server) - Connect to IDP (Config for Azure AD iver.se)
+   - Config instructions documented [here](https://www.ory.sh/hydra/docs/reference/configuration)
+   - [OpenID-Configuration](https://login.microsoftonline.com/common/.well-known/openid-configuration)
+   -
+3. Wharf-web - configure and setup [angular-auth-oicd-client](https://github.com/damienbod/angular-auth-oidc-client)
+   - Implement login endpoint
+   - Implement consent endpoint
+   - Implement logout endpoint
+
+4. Wharf-api - Implement and test
+   - [Go SDK for ory Hydra](https://www.ory.sh/hydra/docs/sdk/go)
+   - Inspiration for [gin middleware here](https://github.com/janekolszak/gin-hydra/blob/master/ginhydra.go)
+     note this is outdated.
```

Comment:
I'm having trouble understanding what ORY Hydra actually solves for us.

I think it would help with a graph (graphical/image or ASCII/Markdown art) that shows where ORY Hydra is meant to live.

1. Does the IDPs, such as Azure AD, talk to ORY Hydra or talk to our API then then talks to ORY Hydra?
2. Does ORY Hydra need any persistence? Do we give it direct access to our DB, or do our API act as a DB proxy, or does it have its own storage?
3. When someone wants to use our API as a service, do they also need the ORY Hydra SDK?

Reply:
Ory Hydra is an implementation of the OpenID Connect standard. The standards documents themselves reccomend using an approved solution to implement the standard. Ory Hydra is one such approved solution.

See below for a map-
![image](https://user-images.githubusercontent.com/4283527/126124274-dad765b8-901a-4da6-a9a7-132750f3e4bf.png)
In this case the providers are just interfaces to the wharf-api and not exposed microservices for simplicity.

1. The wharf's environment link out to IdPs is the ory-hydra pod it acts as a gateway for all things identity.
2. Yes, for the tests i have done thus far i have been using postgres as persistent storage, but other options are available.
3. No but they will need to conform to OpenId Connect.

Reply:
Very nice graphic, and fine answers! Would like to have this explained in the RFC in a nice way as well, so it answers my questions there. (I've not checked yet if you've already added them)